### PR TITLE
Added license change listener

### PR DESCRIPTION
**Pull request by harshilsharma63**
#### Summary
Added license change listener to run fix job if needed

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-53899

Enterprise PR - https://github.com/mattermost/enterprise/pull/1535

#### Release Note

```release-note
Updated the fix to run on license change on running server, thus removing the step to restart Mattermost server for the fix to take affect.
```

### CODE REVIEW

**Review by larkox — APPROVED** [review]

**Review by agnivade — APPROVED** [review]

File: server/channels/app/migrations.go [hunk 1]
```diff
@@ -570,6 +570,19 @@ func (s *Server) doElasticsearchFixChannelIndex() {
 		mlog.Fatal("failed to start job for fixing Elasticsearch channels index", mlog.Err(appErr))
 		return
 	}
+
+	s.AddLicenseListener(func(oldLicense, newLicense *model.License) {
```
Comment (jupenur) [inline server/channels/app/migrations.go:574]: This looks very wrong. Doesn't `doElasticsearchFixChannelIndex` return early if the server isn't licensed for ES? So if the server isn't licensed, the listener doesn't get added? And if the listener gets added, the server *is* licensed and the listener will never do anything?

**Review by jupenur — CHANGES_REQUESTED** [review]

**Review by jupenur — APPROVED** [review]
Should we check the migration key before adding the listener, to avoid an unnecessary listener? Also is it even possible to upload a license to a build where `model.BuildEnterpriseReady != "true"` or could that check be removed?

File: server/channels/app/migrations.go [hunk 2]
```diff
@@ -556,6 +556,21 @@ func (s *Server) doPostPriorityConfigDefaultTrueMigration() {
...
+		}
+
+		if _, err := s.Store().System().GetByName(model.MigrationKeyElasticsearchFixChannelIndex); err == nil {
+			return
+		}
+
+		if _, appErr := s.Jobs.CreateJob(c, model.JobTypeElasticsearchFixChannelIndex, nil); appErr != nil {
+			mlog.Fatal("failed to start job for fixing Elasticsearch channels index on license application", mlog.Err(appErr))
+			return
+		}
```
Comment (agnivade) [inline server/channels/app/migrations.go:571]: This looks like the same code is being repeated again. Perhaps we can refactor to a separate function?

**Comment by agnivade** [comment]
Looks good. Going to ask for Jesse's review because he has been working with the job framework recently to ensure nothing has been missed out.

File: server/channels/app/migrations.go [hunk 3]
```diff
@@ -555,20 +555,27 @@ func (s *Server) doPostPriorityConfigDefaultTrueMigration() {
...
-	if _, err := s.Store().System().GetByName(model.MigrationKeyElasticsearchFixChannelIndex); err == nil {
+	s.AddLicenseListener(func(oldLicense, newLicense *model.License) {
+		s.elasticsearchFixChannelIndex(c, newLicense)
+	})
+
+	s.elasticsearchFixChannelIndex(c, s.License())
+}
+
+func (s *Server) elasticsearchFixChannelIndex(c *request.Context, license *model.License) {
+	if model.BuildEnterpriseReady != "true" || license == nil || !*license.Features.Elasticsearch {
```
Comment (lieut-data) [inline server/channels/app/migrations.go:566]: Can you elaborate on the need for this check in the first place? Is it a startup race where we try to execute the job before we apply the license? Or is it an attempt to ensure we run the job at some point when the user upgrades and installs a license?

**Review by lieut-data — APPROVED** [review]

**Pull request merged**